cpp/tests/fixed_point/fixed_point_tests.cpp

```diff
@@ -0,0 +1,217 @@
+    EXPECT_EQ(1.2345,   num4.get());
+    EXPECT_EQ(1.23456,  num5.get());
+    EXPECT_EQ(1.234567, num6.get());
+
+}
+
+TEST_F(FixedPointTest, SimpleBinaryFPConstruction) {
+
+    using binary_fp32 = fixed_point<int32_t, Radix::BASE_2>;
+
+    binary_fp32 num0{10, scale_type{0}};
+    binary_fp32 num1{10, scale_type{1}};
+    binary_fp32 num2{10, scale_type{2}};
+    binary_fp32 num3{10, scale_type{3}};
+    binary_fp32 num4{10, scale_type{4}};
+
+    EXPECT_EQ(10, num0.get());
+    EXPECT_EQ(10, num1.get());
+    EXPECT_EQ(8,  num2.get());
+    EXPECT_EQ(8,  num3.get());
+    EXPECT_EQ(0,  num4.get());
+
+}
+
+TEST_F(FixedPointTest, SimpleDecimal32Math) {
+
+    decimal32 ONE  {1, scale_type{-2}};
+    decimal32 TWO  {2, scale_type{-2}};
+    decimal32 THREE{3, scale_type{-2}};
+    decimal32 SIX  {6, scale_type{-2}};
+
+    EXPECT_TRUE(ONE + ONE == TWO);
+
+    EXPECT_EQ(ONE   + ONE, TWO);
+    EXPECT_EQ(ONE   * TWO, TWO);
+    EXPECT_EQ(THREE * TWO, SIX);
+    EXPECT_EQ(THREE - TWO, ONE);
+    EXPECT_EQ(TWO   / ONE, TWO);
+    EXPECT_EQ(SIX   / TWO, THREE);
+
+}
+
+TEST_F(FixedPointTest, Decimal32TrickyDivision) {
+
+    decimal32 ONE_1  {1,  scale_type{1}};
+    decimal32 SIX_0  {6,  scale_type{0}};
+    decimal32 SIX_1  {6,  scale_type{1}};
+    decimal32 TEN_0  {10, scale_type{0}};
+    decimal32 TEN_1  {10, scale_type{1}};
+    decimal32 SIXTY_1{60, scale_type{1}};
+
+    decimal32 ZERO = ONE_1;
+
+    EXPECT_EQ(ONE_1.get(),    0); // 1 / 10 = 0
+    EXPECT_EQ(SIX_1.get(),    0); // 6 / 10 = 0
+    EXPECT_EQ(TEN_0.get(),   10);
+    EXPECT_EQ(SIXTY_1.get(), 60);
+
+    EXPECT_EQ(SIXTY_1 / TEN_0, ZERO);
+    EXPECT_EQ(SIXTY_1 / TEN_1, SIX_0);
+
+    decimal32 A{34.56, scale_type{-2}};
+    decimal32 B{1.234, scale_type{-3}};
+    decimal32 C{1,     scale_type{-2}};
+
+    EXPECT_EQ((A / B).get(),       20);
+    EXPECT_EQ(((A * C) / B).get(), 28);
+
+}
+
+TEST_F(FixedPointTest, OverflowDecimal32) {
+
+    #if defined(__CUDACC_DEBUG__)
+
+    decimal32 num0{ 2, scale_type{-9}};
+    decimal32 num1{-2, scale_type{-9}};
+
+    ASSERT_NO_THROW(num0 + num0);
+    ASSERT_NO_THROW(num1 - num0);
+
+    decimal32 min{std::numeric_limits<int32_t>::min(), scale_type{0}};
+    decimal32 max{std::numeric_limits<int32_t>::max(), scale_type{0}};
+    decimal32 NEG_ONE{-1, scale_type{0}};
+    decimal32 ONE{1, scale_type{0}};
+    decimal32 TWO{2, scale_type{0}};
+
+    ASSERT_NO_THROW(min / NEG_ONE);
+    ASSERT_NO_THROW(max * TWO);
+    ASSERT_NO_THROW(min * TWO);
+    ASSERT_NO_THROW(max + ONE);
+    ASSERT_NO_THROW(max - NEG_ONE);
+    ASSERT_NO_THROW(min - ONE);
+    ASSERT_NO_THROW(max - NEG_ONE);
+
+    #endif
+
+}
+
+template<typename ValueType,
+         typename std::enable_if_t<std::numeric_limits<ValueType>::is_integer>* = nullptr>
+void transform_test(std::vector<decimal32> const& vec1,
+                    std::vector<ValueType> const& vec2) {
+
+    std::vector<ValueType> vec3(vec1.size());
+
+    std::transform(
```

Comment:
I'd like to see a thrust::transform test. We need to use these in device code.

Reply:
Added.